drivers/counter/counter_realtek_rts5912_slwtmr.c

```diff
@@ -0,0 +1,344 @@
+
+/**
+ * @file
+ * @brief Realtek RTS5912 Counter driver
+ *
+ * This is the driver for the 32-bit counters on the Realtek SoCs.
+ *
+ * Notes:
+ * - The counters are running in down counting mode.
+ * - Interrupts are triggered (if enabled) when the counter
+ *   reaches zero.
+ * - These are not free running counters where there are separate
+ *   compare values for interrupts. When setting single shot alarms,
+ *   the counter values are changed so that interrupts are triggered
+ *   when the counters reach zero.
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/counter.h>
+#include <soc.h>
+#include <errno.h>
+#include <stdbool.h>
+#include "reg/reg_slwtmr.h"
+#include "reg/reg_system.h"
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_rts5912.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(counter_realtek_rts5912_slwtmr, CONFIG_COUNTER_LOG_LEVEL);
+
+struct counter_rts5912_config {
+	struct counter_config_info info;
+	void (*config_func)(void);
+	uint32_t base_address;
+	uint16_t prescaler;
+#if (CONFIG_CLOCK_CONTROL)
+	const struct device *clk_dev;
+	struct rts5912_sccon_subsys sccon_cfg;
+#endif
+};
+
+struct counter_rts5912_data {
+	counter_alarm_callback_t alarm_cb;
+	counter_top_callback_t top_cb;
+	void *user_data;
+};
+
+#define COUNTER_RTS5912_REG_BASE(_dev)                                                             \
+	((SLWTMR_Type *)((const struct counter_rts5912_config *const)_dev->config)->base_address)
+
+#define COUNTER_RTS5912_CONFIG(_dev) (((const struct counter_rts5912_config *const)_dev->config))
+
+#define COUNTER_RTS5912_DATA(_dev) ((struct counter_rts5912_data *)dev->data)
+
+static int counter_rts5912_start(const struct device *dev)
+{
+	SLWTMR_Type *counter = COUNTER_RTS5912_REG_BASE(dev);
+
+	if (counter->CTRL & SLWTMR_CTRL_EN) {
+		return -EALREADY;
+	}
+
+	counter->CTRL |= SLWTMR_CTRL_EN;
+
+	LOG_DBG("%p Counter started", dev);
+
+	return 0;
+}
+
+static int counter_rts5912_stop(const struct device *dev)
+{
+	SLWTMR_Type *counter = COUNTER_RTS5912_REG_BASE(dev);
+
+	if (!(counter->CTRL & SLWTMR_CTRL_EN)) {
+		/* Already stopped, nothing to do */
+		return 0;
+	}
+
+	/* disable timer and disable interrupt. */
+	counter->CTRL = 0;
+	counter->LDCNT = counter->CNT;
+
+	/* w1c interrupt pending status */
+	counter->INTSTS |= SLWTMR_INTSTS_STS;
+
+	LOG_DBG("%p Counter stopped", dev);
+
+	return 0;
+}
+
+static int counter_rts5912_get_value(const struct device *dev, uint32_t *ticks)
+{
+	SLWTMR_Type *counter = COUNTER_RTS5912_REG_BASE(dev);
+
+	*ticks = counter->CNT;
+	return 0;
+}
+
+static int counter_rts5912_set_alarm(const struct device *dev, uint8_t chan_id,
+				     const struct counter_alarm_cfg *alarm_cfg)
+{
+	SLWTMR_Type *counter = COUNTER_RTS5912_REG_BASE(dev);
+	struct counter_rts5912_data *data = COUNTER_RTS5912_DATA(dev);
+	const struct counter_rts5912_config *counter_cfg = COUNTER_RTS5912_CONFIG(dev);
```

Comment:
You don't need to cast to get the driver `data` and `config` structures.  Please remove all the COUNTER_RTS5912_DATA and COUNTER_RTS5912_CONFIG macros.
```suggestion
	struct counter_rts5912_data *data = dev->data;
	const struct counter_rts5912_config *counter_cfg = dev->config;
```

Reply:
done